### PR TITLE
Add MemoryExtra in InterpretCx constructor params

### PR DESCRIPTION
**Pull request by pvdrz**
This is to avoid modifying `MemoryExtra` inside `InterpretCx` after initialization. Related miri PR: https://github.com/rust-lang/miri/pull/792

r? @RalfJung

### CODE REVIEW

**Comment by RalfJung** [comment]
Please also change `Machine` trait in `machine.rs` to not require a `Default` bound for `MemoryExtra`.

**Comment by christianpoveda** [comment]
> Please also change `Machine` trait in `machine.rs` to not require a `Default` bound for `MemoryExtra`.

I believe this is still required for all the `Memory`  initializations inside the compiler. Should I remove it then?

**Comment by RalfJung** [comment]
That code will still work fine because there we are using a concrete `MemoryExtra` that *does* implement `Default`.

But what the trait says is that anyone implementing the `Machine` interface must provide a `MemoryExtra` that implements `Default`. There is no reason left to require that.

File: src/librustc_mir/const_eval.rs [hunk 1]
```diff
@@ -632,7 +632,11 @@ pub fn const_eval_raw_provider<'tcx>(
     }
 
     let span = tcx.def_span(cid.instance.def_id());
-    let mut ecx = InterpretCx::new(tcx.at(span), key.param_env, CompileTimeInterpreter::new());
+    let mut ecx = InterpretCx::new(
+        tcx.at(span),
+        key.param_env,
+        CompileTimeInterpreter::new(),
+        Default::default());
```
Comment (RalfJung) [inline src/librustc_mir/const_eval.rs]: Please follow rustfmt: The final `)` should be on its own line, and less indented.

**Review by RalfJung — APPROVED** [review]
This looks great! r=me once Travis is green and with the nit fixed.

@bors delegate

File: src/librustc_mir/const_eval.rs (hunk 1 above)
Comment (christianpoveda) [reply src/librustc_mir/const_eval.rs]: Oh ok sorry, I don't know why my text editor is messing the identation.

Comment (RalfJung) [reply src/librustc_mir/const_eval.rs]: It didn't, the parenthesis wasn't even on its own line. ;)

**Comment by christianpoveda** [comment]
r=@RalfJung

**Comment by RalfJung** [comment]
It would be `@bors r=RalfJung`. Note the lack of a second `@`. ;)

@bors r+

**Comment by christianpoveda** [comment]
> It would be `@ bors r=RalfJung`. Note the lack of a second `@`. ;)
> 
> @bors r+

Oh ok! Thank you!

**Comment by christianpoveda** [comment]
Remember that https://github.com/rust-lang/rust/pull/62158 needs these changes to work

**Comment by RalfJung** [comment]
I do. The thing is, we are about to branch off beta in less than a week, which means changes that break Miri cannot land. So once https://github.com/rust-lang/rust/pull/62105 lands, we'll have to fold a Miri update into this PR.

**Comment by RalfJung** [comment]
Looks very much like the Miri update will land first, so let's avoid wasting CI time.

@bors r-

**Comment by RalfJung** [comment]
@christianpoveda okay the Miri PR landed, are you ready to update this one? I prepared the `rustup` branch in the Miri repository for you. The sequence of commands for you (assuming you checked out your local branch for this PR) should roughly be
```
git fetch origin
git rebase origin/master # this should just work without conflicts
git status # this should show a clean working dir!
cd src/tools/miri
git fetch origin
git reset --hard origin/rustup
cd ../../..
./x.py # updates Cargo.lock if needed
git commit -am "update miri"
```

**Comment by RalfJung** [comment]
Looks great. :)

@bors r+

**Comment by RalfJung** [comment]
Now how do I get rid of this "rollup=never" thing...

@bors rollup-

**Comment by RalfJung** [comment]
@christianpoveda we got overtaken by another commit that also modifies Miri... can you do a rebase and do the submodule update again? I updated the `rustup` branch on the Miri side.

**Comment by RalfJung** [comment]
Alternatively, you can remove the miri-updating commit and we wait until Wednesday, when we get out of the "freeze" for tools.

**Comment by christianpoveda** [comment]
> Alternatively, you can remove the miri-updating commit and we wait until Wednesday, when we get out of the "freeze" for tools.

I will do this, it seems the better option right now

**Comment by RalfJung** [comment]
It's just 4h until it's Wednesday UTC, I feel confident this will not land fast enough to cause problems and I want to stop having to remember.^^

@bors r+

**Pull request merged**